### PR TITLE
Add `internal_features` lint

### PR DESCRIPTION
**Pull request by Noratrieb**
Implements https://github.com/rust-lang/compiler-team/issues/596

Also requires some more test blessing for codegen tests etc

@jyn514 had the idea of just `allow`ing the lint by default in the test suite. I'm not sure whether this is a good idea, but it's definitely one worth considering. Additional input encouraged.

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -214,19 +240,19 @@ declare_features! (
...
     /// Allows using the `#![panic_runtime]` attribute.
-    (active, panic_runtime, "1.10.0", Some(32837), None),
+    (internal, panic_runtime, "1.10.0", Some(32837), None),
     /// Allows using `#[rustc_allow_const_fn_unstable]`.
     /// This is an attribute on `const fn` for the same
     /// purpose as `#[allow_internal_unstable]`.
-    (active, rustc_allow_const_fn_unstable, "1.49.0", Some(69399), None),
+    (internal, rustc_allow_const_fn_unstable, "1.49.0", Some(69399), None),
     /// Allows using compiler's own crates.
     (active, rustc_private, "1.0.0", Some(27812), None),
```
Comment (jyn514) [inline compiler/rustc_feature/src/active.rs:253]: Shouldn't this one be internal too? Or are you counting rustc_private as a public api?

File: compiler/rustc_lint/locales/en-US.ftl [hunk 2]
```diff
@@ -442,6 +442,9 @@ lint_builtin_incomplete_features = the feature `{$name}` is incomplete and may n
     .note = see issue #{$n} <https://github.com/rust-lang/rust/issues/{$n}> for more information
     .help = consider using `min_{$name}` instead, which is more stable and complete
 
+lint_builtin_internal_features = the feature `{$name}` is interal to the compiler or standard library
```
Comment (jyn514) [inline compiler/rustc_lint/locales/en-US.ftl:445]:
```suggestion
lint_builtin_internal_features = the feature `{$name}` is internal to the compiler or standard library
```

**Review by jyn514 — COMMENTED** [review]
❤️ ❤️

File: compiler/rustc_feature/src/active.rs (hunk 1 above)
Comment (Nilstrieb) [reply compiler/rustc_feature/src/active.rs:253]: I wouldn't want to emit a "you're not supposed to use this" error for this, people _are_ supposed to use this to write cool tools!

File: tests/ui/abi/unsupported.rs [hunk 3]
```diff
@@ -8,6 +8,8 @@
 // [aarch64] compile-flags: --target=aarch64-unknown-linux-gnu --crate-type=rlib
 // [arm] needs-llvm-components: arm
 // [arm] compile-flags: --target=armv7-unknown-linux-gnueabihf --crate-type=rlib
+
+#![allow(internal_features)]
```
Comment (oli-obk) [inline tests/ui/abi/unsupported.rs:12]: oof. How about just allowing these on all tests like we do with the `dead_code` lints?

**Comment by oli-obk** [comment]
Does this affect any tools or backends? (modulo their tests, I just mean the tool/backend code)

**Comment by Nilstrieb** [comment]
From a quick skim over them the tools and backends look fine as they don't use any features I classified as internal, but I'll make sure that they're fine when marking this as ready.

File: src/bootstrap/builder.rs [hunk 4]
```diff
@@ -1933,6 +1933,11 @@ impl<'a> Builder<'a> {
                 rustflags.arg("-Zvalidate-mir");
                 rustflags.arg(&format!("-Zmir-opt-level={}", mir_opt_level));
             }
+
+            if stage != 0 {
```
Comment (jyn514) [inline src/bootstrap/builder.rs:1937]:
```suggestion
            // cfg(not(bootstrap))
            if stage != 0 {
```
this helps t-release find it when they do the beta bump.

**Comment by Nilstrieb** [comment]
welp, miri's fancy sysroot building stuff doesn't pass the allow flag when building std: https://github.com/rust-lang/rust/actions/runs/4462337928/jobs/7836803005?pr=108955#step:26:3429

@jyn514 recommended I don't do it at first but I guess I really do have to make a stdarch PR first with an `allow(unknown_lints)`

**Comment by Nilstrieb** [comment]
Status: @oli-obk asked me to make this lint unstable to avoid stabilizing this lint needlessly (since it's not like anyone using it will ever be on stable).

**Comment by Nilstrieb** [comment]
I've tried adding it as an unstable lint but that didn't really work because it's allowed by default in the test suite so it complains about missing features - complicated.
So I propose adding it as a stable lint. I don't see a problem with that - `incomplete_features` are stable too and really, lint's aren't "stable", they're just usable on stable. The lint could be removed at any point, users would just get warnings. For this lint it's especially true because the only time it will be used is when people already enable internal features - getting extra warnings because the lint is removed is not really a concern for them.

The `internal_features` lint lints on any code enabling features that are considered internal to the compiler and standard library. It's deny-by-default. These features include, but are not limited to, `lang_items` and `intrinsics`.

```rust
#![feature(lang_items)]
fn main() {}
```
would be an error by default. Users have to `allow(internal_features)` to allow the lint and let the build pass. The [compiler MCP ](https://github.com/rust-lang/compiler-team/issues/596) for this lint was accepted already.

@rust-lang/lang can you start an FCP? (I think lints are T-lang, right?)

**Comment by scottmcm** [comment]
I was pondering whether this is more lang or more compiler, but with a compiler MCP already complete,

@rfcbot fcp merge

I support giving stronger discouragement from using `S-tracking-perma-unstable` things.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by pnkfelix** [comment]
@scottmcm [wrote](https://github.com/rust-lang/rust/pull/108955#issuecomment-1564739001):

> I was pondering whether this is more lang or more compiler, but with a compiler MCP already complete, [...]

arguably the compiler-team MCP was written in a way that was about adding an experimental unstable lint, while this PR is now adding it as a "stable lint", while it also argues that "stable" doesn't mean the same thing for lints that it does for other features ...

anyway, I think this is overall a good idea.

I'm just trying to figure out whether it should also have gone through T-compiler FCP as well, or if the MCP alone suffices even with upgrade to "stable lint" as described here.

**Comment by pnkfelix** [comment]
I'll bring it up at T-compiler triage tomorrow, find out if the people there thing a T-compiler FCP is warranted or not

**Comment by pnkfelix** [comment]
@rfcbot concern does-this-now-need-a-compiler-team-fcp-as-well

The text of https://github.com/rust-lang/compiler-team/issues/596 that was seconded and had no objections from T-compiler was arguably about an experimental lint, not an insta-stable one. (Although one might argue that the comparison with `incomplete_features` means that we should always have expected this to be in stable, analogous to how that lint is also in stable, for reasons I do not claim to  understand fully.)

Anyways, lets see if anyone at the thursday triage meeting wants to argue that this should go through T-compiler FCP as well.

**Comment by compiler-errors** [comment]
I personally do not think this warrants a compiler FCP. Any use of this lint automatically implies they're using internal features, which means that regular stability guarantees are already out the window, no?

**Comment by tmandry** [comment]
> Any use of this lint automatically implies they're using internal features, which means that regular stability guarantees are already out the window, no?

Very pedantically, one could now write `#![allow(internal_features)]` and not use any unstable features, while also not writing `#![allow(unknown_lints)]`. So in that very odd edge case, removal would break working code.

In any case, I agree with this going forward.

@rfcbot reviewed

**Comment by Nilstrieb** [comment]
Even more pedantically, it wouldn't break code because unknown lints are just a warning. But yeah.

**Comment by oli-obk** [comment]
@rfcbot resolve does-this-now-need-a-compiler-team-fcp-as-well

**Comment by oli-obk** [comment]
@rfcbot concern does-this-now-need-a-compiler-team-fcp-as-well

**Comment by pnkfelix** [comment]
@rfcbot resolve does-this-now-need-a-compiler-team-fcp-as-well

**Comment by pnkfelix** [comment]
@rfcbot reviewed

**Comment by Nilstrieb** [comment]
@compiler-errors found that this is a legitimate clippy bug, it passes the wrong substs
https://github.com/rust-lang/rust/blob/32d81eccd64513bacef9dfa1574543ada6b45d85/src/tools/clippy/clippy_lints/src/incorrect_impls.rs#L85
The debug assertions rightfully triggers in rustc.
..how did this land on master and why is it erroring now?? @flip1995 do you have any idea?

**Comment by flip1995** [comment]
> ..how did this land on master and why is it erroring now?? @flip1995 do you have any idea?

I don't know... I don't go over the thousands of lines of Clippy changes when syncing to rustc repo. So I only catch what `./x.py test` shows me.

File: compiler/rustc_feature/src/active.rs [hunk 5]
```diff
@@ -137,52 +163,52 @@ declare_features! (
...
-    (active, allow_internal_unstable, "1.0.0", None, None),
+    (internal, allow_internal_unstable, "1.0.0", None, None),
     /// Allows using anonymous lifetimes in argument-position impl-trait.
     (active, anonymous_lifetime_in_impl_trait, "1.63.0", None, None),
     /// Allows identifying the `compiler_builtins` crate.
     (active, compiler_builtins, "1.13.0", None, None),
     /// Allows writing custom MIR
-    (active, custom_mir, "1.65.0", None, None),
+    (internal, custom_mir, "1.65.0", None, None),
     /// Outputs useful `assert!` messages
```
Comment (matthiaskrgr) [inline compiler/rustc_feature/src/active.rs:183]: shouldn't `compiler_builtins` be internal as well?

Comment (Nilstrieb) [reply compiler/rustc_feature/src/active.rs:183]: yeah there are many that could be internal too, I was mostly concerned with the ones that will cause ICEs if used incorectly
I added it too. But I don't think it's important to get this entirely right on this PR, we can always change things later. The most important thing is that lang_items, intrinsics, custom_mir and friends are marked as internal.

**Comment by Nilstrieb** [comment]
@bors rollup=iffy
my hopes that it passes first time are low

**Comment by Nilstrieb** [comment]
Blocked on the next cargo sync

**Comment by weihanglo** [comment]
Cargo update is live now.
<https://github.com/rust-lang/rust/pull/114364>

**Comment by Nilstrieb** [comment]
rebased (although that's technically unnecessary as the merge commit is tested)
@bors r=oli-obk

**Pull request merged**